### DISCUSSION TITLE
cannot work well while using LocalExecutor or CeleryExecutor

### DISCUSSION THREAD

**Discussion by yijianxilai** [apache/airflow#18489] [Q&A]
setting: sql_alchemy_conn = sqlite:////root/airflow/airflow.db,
dags and UI can run when Executor is SequentialExecutor,  but very serious error occured in  LocalExecutor or CeleryExecutor mode.
UI task log is : 
*** Log file does not exist: /root/airflow/logs/a_pipeline_warehouse_29/load_data_to_dwd/2021-09-22T00:00:00+00:00/2.log
*** Fetching from: http://:8793/log/a_pipeline_warehouse_29/load_data_to_dwd/2021-09-22T00:00:00+00:00/2.log
*** Failed to fetch log file from worker. The request to ':///' is missing either an 'http://'                         or 'https://' protocol.

**Comment by potiuk** [comment]
sqlite should only work with Sequential Executor

**Reply by potiuk** [reply]
Don't use sqlite for anything else than testing and development. When you start Airflow (new vesions) it will even refuse to work with other executors and you get clear warning in the UI that you should use it only for development/testing

**Reply by dstandish** [reply]
@potiuk i know that the scheduler & executors have changed a lot over the years...  do you know if disabling sqlite with localexecutor is still valid?  e.g. i just tried commenting out this line:

```python
        # if is_sqlite and is_executor_without_sqlite_support:
        #     raise AirflowConfigException(f"error: cannot use sqlite with the {self.get('core', 'executor')}")
```

and running this dag:

```python
from airflow.models.dag import DAG
from airflow.operators.dummy import DummyOperator
from airflow.operators.python import PythonOperator
import pendulum

with DAG(dag_id='hello', start_date=pendulum.now().add(days=-5)) as dag:
    for num in range(100):
        o1 = PythonOperator(
            task_id=f'hello_{num}_a',
            start_date=pendulum.now().add(days=-5),
            python_callable=lambda: print('hello'),
        )
        o2 = PythonOperator(
            task_id=f'hello_{num}_b',
            start_date=pendulum.now().add(days=-5),
            python_callable=lambda: print('hello'),
        )
        o1 >> o2
```

And it seemed to work fine!

I know that sqlite is not the production choice, but for dev and testing, having parallelism while using sqlite could be convenient, and if it actually does work, it could be a better experience for new adopters (since dags with many tasks will run faster).  in my test of localexecutor, around 400 tasks completed in around 2 minutes.  trying it again using sequential it is looking like it will be roughly 15 minutes.

in principle it doesn't seem that having no concurrency in the db should necessarily prevent concurrent execution of tasks...

**Reply by potiuk** [reply]
Hey @dstandish  - Indeed - currently no parallelism is possible in Sqlite, and while theorethically some level of parallelism can be achieved, I do not think we should do it.

Sqlite cannot handle multiple parallel writes to the database and Airflow is build around the concept of having many parallell processes writing to DB at the same time - and this is where it gets the "concurrency" speed-ups of scheduler when you are talking about parallel execution.

From: https://stackoverflow.com/questions/4060772/sqlite-concurrent-access
> SQLite has a readers-writer lock on the database level. Multiple connections (possibly owned by different processes) can read data from the same database at the same time, but only one can write to the database.

Theorethically - I agree - you would not need it all the time.  When you have long running tasks in executor, the database lock is not needed, so in theory you COULD have multiple processes competing for write access and it would speed up things. However when you pair it with scheduler also writing to the database, then you have huge potential for deadlocks and solving them might be next to impossible or at the least - very difficult.

Also I'd say this is a "GREAT" property that sqlite is so much slower when it comes to parallelism. For multiple reasons using SQlite as production database for Airflow is a bad idea (backup, resiliency, high availability etc.). So you should not do it. The fact that it is so slow, makes it even impossible to use in production - which is actually very good.

If we try to make it run in parallel - we risk that some people might be tempted to use it in production, which is something we definitely do not want to support. So we have no incentive as community to invest in parallelism of sqlite. If we do - we will have harder time in handling issues of people who will try (and fail) running it in production because it will be faster. 

Right now in latest version of Airflow you will even have a "notice" that you should not use sqlite in production displayed at the top of the UI. Because we want to make people aware they shoudl not use it. 

I'd say we will never enable, endorse, nor support parallelism in sqlite. Especially that this is super easy to run your development environment in Postgres - for example using docker-compose (see https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html for example) which gives you nice local development start which you can even adapt to your needs as you see fit - you will get all the performance resulting from parallelism there.

Sqlite will remain as a great way to say - run unit tests with Airflow. But nothing more.

This is my own opinion, of course not a general truth. As everything in Airflow - it is all decided on the community level. If you have some arguments and would like to promote sqlite to more production-like setup - feel free to start a thread in devlist and  make your arguments there. Maybe you will manage to convince the community that it is worth it (and maybe even you could take on the task of doing it if you get to a consensus there)

**Reply by dstandish** [reply]
Thanks Jarek

Yeah definitely wasn't suggesting that we promote sqlite to a supported production database... Just wondering, for the local dev scenario, whether that prohibition is still warranted.

It's true it's not that hard to install postgres or mysql (even if you want to stay away from docker).

But yeah I think the bottom line is that, probably even to prove that it's worth changing it from a strict prohibition perhaps to a stern warning would take substantial effort, and there may be higher value things to work on.

**Reply by potiuk** [reply]
I think it's all about psychology and managing expectations really. It's not even a technical problem. There are simply two kinds of users - those who read docs and those who don't. I am now thinking about the latter.

We have to REALLY be careful what we give to people as "community". A good example is "quiick-start docker-compose" which is officially published on our page. We have a lot of disclaimers there "do not use it for production inside the file", a warning in the documentation which explicitly states "this is not production ready and requires a lot of work". Yet we continue having people asking "We used YOUR OFFICIAL docker-compose for deployment and it does not work. How do I fix it"? Often people don't read/check what they can/cannot do with the software, they just try, and when it works for their try, they might jump to conclusions and make assumptions that later will lead them to unrealistic expectations and creating issues that things "do not work as they expect".

Again - I think this is all about managing expectations - we have to be very explicit what we give to our users and what can they expect from us. If we work on making sqlite works "fast" I think there is no way to avoid people starting using it in production. I think it's actually better if it slow (and appropriate warning displayed in the UI explaining it), because then it will not give false expectation to the users who do not read the documentation (and they will realize their expectations are unrealistic much earlier).

**Reply by dstandish** [reply]
> Yet we continue having people asking "We used YOUR OFFICIAL docker-compose for deployment and it does not work. How do I fix it"?

🙂
Yeah, totally hear you on being very selective and deliberate about what you put out there.  Makes complete sense.  Disclaimers have limited value because people are in general in a hurry and just trying to get things working and only will go deeper (e.g. into the docs) when necessary.  Also, warnings like "this is not meant for production" have a particular issue because as a user you're like... well why not?  Often times we don't really explain those kind of statements in that great of detail, at least as far as I recall.  So as a user you're like "well it seems to work locally and it's probably good enough for me"
